lib/go-atscfg/remapdotconfig.go

```diff
@@ -0,0 +1,413 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"errors"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+)
+
+const CacheURLParameterConfigFile = "cacheurl.config"
+const CacheKeyParameterConfigFile = "cachekey.config"
+
+type RemapConfigDSData struct {
+	ID                       int
+	Type                     tc.DSType
+	OriginFQDN               *string
+	MidHeaderRewrite         *string
+	CacheURL                 *string
+	RangeRequestHandling     *int
+	CacheKeyConfigParams     map[string]string
+	RemapText                *string
+	EdgeHeaderRewrite        *string
+	SigningAlgorithm         *string
+	Name                     string
+	QStringIgnore            *int
+	RegexRemap               *string
+	FQPacingRate             *int
+	DSCP                     int
+	RoutingName              *string
+	MultiSiteOrigin          *string
+	Pattern                  *string
+	RegexType                *string
+	Domain                   *string
+	RegexSetNumber           *string
+	OriginShield             *string
+	ProfileID                *int
+	Protocol                 *int
+	AnonymousBlockingEnabled *bool
+	Active                   bool
+}
+
+func MakeRemapDotConfig(
+	serverName tc.CacheName,
+	toToolName string, // tm.toolname global parameter (TODO: cache itself?)
+	toURL string, // tm.url global parameter (TODO: cache itself?)
+	atsMajorVersion int,
+	cacheURLConfigParams map[string]string, // map[name]value for this server's profile and config file 'cacheurl.config'
+	dsProfilesCacheKeyConfigParams map[int]map[string]string, // map[profileID][paramName]paramVal for this server's profile, config file 'cachekey.config'
+	serverPackageParamData map[string]string, // map[paramName]paramVal for this server, config file 'package'
+	serverInfo *ServerInfo, // ServerInfo for this server
+	remapDSData []RemapConfigDSData,
+) string {
+	hdr := GenericHeaderComment(string(serverName), toToolName, toURL)
+	text := ""
+	if tc.CacheTypeFromString(serverInfo.Type) == tc.CacheTypeMid {
+		text = GetServerConfigRemapDotConfigForMid(atsMajorVersion, dsProfilesCacheKeyConfigParams, serverInfo, remapDSData, hdr)
+	} else {
+		text = GetServerConfigRemapDotConfigForEdge(cacheURLConfigParams, dsProfilesCacheKeyConfigParams, serverPackageParamData, serverInfo, remapDSData, atsMajorVersion, hdr)
+	}
+	return text
```

Comment:
Does this spit out trailing newlines? I don't think it does.

Reply:
actually, after looking through the rest of it, I think it might.

Reply:
Yes.